### ISSUE TITLE
:keyword_list and :nonempty_keyword_list should apply nested options defaults

### ISSUE THREAD

**Issue by whatyouhide** [dashbitco/nimble_options#51]
Today, I have this schema:

```elixir
schema = [
  pool_options: [
    type: :keyword_list,
    default: [],
    keys: [
      protocol: [type: :atom, default: :http1],
      size: [type: :integer, default: 10]
    ]
  ]
]
```

However, if I validate `[]` I get this:

```elixir
NimbleOptions.validate!([], schema)
#=> [pool_options: []]
```

when in reality I'd expect that the nested defaults are propagated up, having the **expected behavior** be:

```elixir
NimbleOptions.validate!([], schema)
#=> [pool_options: [protocol: :http1, size: 10]]
```

@josevalim @msaraiva I can work on this if we agree it's the right call. Thoughts?

**Comment by wojtekmach** [comment]
+1 from me.

Btw, I think in this particular case the `default: []` is misleading as that's not what we expect to happen, maybe we should warn or document it?

Worth mentioning that today we could set it to `default: [protocol: :http1, size: 10]` but it doesn't compose well, with the same schema as above we have:

```elixir
iex> NimbleOptions.validate!([pool_options: [protocol: :http2]], schema)
[pool_options: [protocol: :http2]] # (no port!)
```

**Comment by msaraiva** [comment]
@whatyouhide are you using the version on master? I believe this has been addressed in #50. More details can be found here https://github.com/dashbitco/nimble_options/issues/49#issuecomment-698406926.

Using master I get:

```
[
  pool_options: [
    type: :keyword_list,
    default: [],
    keys: [
      protocol: [type: :atom, default: :http1],
      size: [type: :integer, default: 10] 
    ]
  ]
]
NimbleOptions.validate!([], schema)
#=> [pool_options: [size: 10, protocol: :http1]]
```

**Comment by whatyouhide** [comment]
@msaraiva is right, this is fixed in master :)
